src/logic/map_objects/descriptions.cc

```diff
@@ -422,9 +431,26 @@ void Descriptions::add_object_description(const LuaTable& table, MapObjectType t
 	case MapObjectType::RESOURCE:
 		resources_->add(new ResourceDescription(table));
 		break;
-	case MapObjectType::TERRAIN:
-		terrains_->add(new TerrainDescription(table, *this));
-		break;
+	case MapObjectType::TERRAIN: {
+		const DescriptionManager::RegistryCallerInfo& c =
+		   description_manager_->get_registry_caller_info(type_name);
+		assert(c.second.empty() ^ (c.first == DescriptionManager::RegistryCallerType::kWorldAddon));
+		uint32_t index = 0;
+		if (!c.second.empty()) {
+			index = addons_.size();
+			for (uint32_t i = 0; i < addons_.size(); ++i) {
+				if (addons_[i]->internal_name == c.second) {
+					index = i;
+					break;
+				}
+			}
+			if (index >= addons_.size()) {
+				throw wexception("Terrain %s defined by add-on %s which is not enabled",
+				                 type_name.c_str(), c.second.c_str());
+			}
+		}
+		terrains_->add(new TerrainDescription(table, *this, index));
```

Comment:
You should at +1 to the index, otherwise the layer from the very first addon might conflict with official layers.